criu/files-reg.c

```diff
@@ -168,52 +170,118 @@ static int copy_chunk_from_file(int fd, int img, off_t off, size_t len)
 
 	while (len > 0) {
 		ret = sendfile(img, fd, &off, len);
-		if (ret <= 0) {
+		if (ret < 0) {
 			pr_perror("Can't send ghost to image");
 			return -1;
 		}
+		if (ret == 0) {
```

Comment:
We know exact size that we want to copy and so we have to return an error if sendfail return 0. Do I miss something?

Reply:
Fiemap returns page-alligned extents, so if the extent is only partially initialized send file can return 0 in the middle. Example:

```
echo test > testfile

filefrag -v testfile
Filesystem type is: 9123683e
File size of testfile is 5 (1 block of 4096 bytes)
 ext:     logical_offset:        physical_offset: length:   expected: flags:
   0:        0..    4095:          0..      4095:   4096:             last,not_aligned,inline,eof
testfile: 1 extent found

xfs_io -c 'sendfile -i testfile 0 4096' testcopy
sent 5/4096 bytes at offset 0
5,000000 bytes, 1 ops; 0.0000 sec (180,845 KiB/sec and 37037,0370 ops/sec)

xfs_io -c 'sendfile -i testfile 5 4091' testcopy
sent 0/4091 bytes at offset 5
0,000000 bytes, 0 ops; 0.0000 sec (0,000000 bytes/sec and 0,0000 ops/sec)
```

So for convenience of passing fiemap extents directly to sendfile we need to handle 0 return as success even is len is not zero.

@featherchen  please add this example to your patch commit message.

Reply:
Thanks, I have added it to the commit message, please check.

Reply:
@Snorch it returns 0 only if an uninitialized extent is beyond the file size, doesn't it?

Reply:
@avagin Yes, and we call sendfile on uninitialized extent beyond the file size due to 1) fiemap 4k granularity 2) copy_chunk_from_file retry logic when len is yet not 0